rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -218,18 +195,75 @@ std::string Recorder::serialized_offered_qos_profiles_for_topic(const std::strin
   return YAML::Dump(offered_qos_profiles);
 }
 
-rclcpp::QoS Recorder::common_qos_or_fallback(const std::string & topic_name)
+rclcpp::QoS Recorder::adapt_qos_to_publishers(const std::string & topic_name) const
 {
   auto endpoints = node_->get_publishers_info_by_topic(topic_name);
-  if (!endpoints.empty() && all_qos_same(endpoints)) {
-    return Rosbag2QoS(endpoints[0].qos_profile()).default_history();
+  size_t num_endpoints = endpoints.size();
+  size_t reliability_reliable_endpoints_count = 0;
+  size_t durability_transient_local_endpoints_count = 0;
+  for (const auto & endpoint : endpoints) {
+    const auto & profile = endpoint.qos_profile().get_rmw_qos_profile();
+    if (profile.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE) {
+      reliability_reliable_endpoints_count++;
+    }
+    if (profile.durability == RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL) {
+      durability_transient_local_endpoints_count++;
+    }
+  }
+
+  // We set policies in order as defined in rmw_qos_profile_t
+  Rosbag2QoS request_qos;
+  // Policy: history, depth
+  // History does not affect compatibility
+  request_qos.default_history();
+
+  // Policy: reliability
+  if (reliability_reliable_endpoints_count == num_endpoints) {
+    request_qos.reliable();
+  } else {
+    if (reliability_reliable_endpoints_count > 0) {
+      ROSBAG2_TRANSPORT_LOG_WARN_STREAM(
+        "Some, but not all, publishers on topic \"" << topic_name << "\" "
+          "are offering Reliable reliability. "
```

Comment:
```suggestion
          "are offering a RELIABLE reliability QoS policy. "
```
Wondering how we can make this helpful in the context of other ros2 nodes running. I'd prefer to be more verbose.

Reply:
Ditto with logs below

Reply:
I don't think we can be _much_ more verbose, it's pretty wordy as it is. I updated the setting names to use exact rmw enum values, for extra specificity.

We do have the node names publishing, but for now i'd rather not print quite so much. The user has the ability to it down via `ros2 topic info`.